tests/common/devices.py

```diff
@@ -554,60 +554,79 @@ def get_version(self):
         output = dut.command("sonic-cfggen -y /etc/sonic/sonic_version.yml -v build_version")
         return output["stdout_lines"][0].strip()
 
+
 class EosHost(AnsibleHostBase):
     """
     @summary: Class for Eos switch
 
     For running ansible module on the Eos switch
     """
 
-    def __init__(self, ansible_adhoc, hostname, user, passwd, gather_facts=False):
-        AnsibleHostBase.__init__(self, ansible_adhoc, hostname, connection="network_cli")
-        evars = { 'ansible_connection':'network_cli', \
-                  'ansible_network_os':'eos', \
-                  'ansible_user': user, \
-                  'ansible_password': passwd, \
-                  'ansible_ssh_user': user, \
-                  'ansible_ssh_pass': passwd, \
-                  'ansible_become_method': 'enable' }
+    def __init__(self, ansible_adhoc, hostname, net_user, net_passwd, shell_user=None, shell_passwd=None, gather_facts=False):
+        self.net_user = net_user
+        self.net_passwd = net_passwd
+        self.shell_user = shell_user
+        self.shell_passwd = shell_passwd
+        AnsibleHostBase.__init__(self, ansible_adhoc, hostname)
+
+    def __getattr__(self, item):
```

Comment:
Should this be __getattribute__? I think you want to setup credential for all methods, right?

Did you try to call both EOS and Linux method from a same instance?

Reply:
I think the idea here is create NetworkCli based host if user call eos module and create LinuxShell based host if user call other module.
The judgement happens when invoking a module because we couldn't keep two types hosts in a single instance and that's why Xin use __getattr__.
The only overhead is we need to override extra-var and create a host every time when calling a module.

Reply:
Thanks @Blueve for the explanation. 
@yxieca According to https://stackoverflow.com/questions/3278077/difference-between-getattr-vs-getattribute:

>A key difference between __getattr__ and __getattribute__ is that __getattr__ is only invoked if the attribute wasn't found the usual ways.

When an attribute is not found on the EosHost or FanoutHost instance, the __getattr__ will try to get a function for running the ansible module. This is the key technique for us to invoke ansible module using format like `fanouthost.<ansible_module_name>(*args, **kwargs)` or `eoshost.<ansible_module_name>(*args, **kwargs)`

Here the ansible module can a EOS module like eos_command, eos_config, or a regular module like command, shell, etc.

Yes, I have tested that both EOS and Linux method can be called from a same instance.